libcudacxx/include/cuda/__mdspan/optimized_accessors.h

```diff
@@ -0,0 +1,184 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of libcu++, the C++ Standard Library for your entire system,
+// under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+// SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _CUDA__MDSPAN_OPTIMIZED_ACCESSORS_H
+#define _CUDA__MDSPAN_OPTIMIZED_ACCESSORS_H
+
+#include <cuda/std/detail/__config>
+
+#if defined(_CCCL_IMPLICIT_SYSTEM_HEADER_GCC)
+#  pragma GCC system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_CLANG)
+#  pragma clang system_header
+#elif defined(_CCCL_IMPLICIT_SYSTEM_HEADER_MSVC)
+#  pragma system_header
+#endif // no system header
+
+#if _CCCL_STD_VER >= 2014
+
+#  include <cub/thread/thread_load.cuh>
+#  include <cub/thread/thread_store.cuh>
+
+#  include <cuda/std/__cccl/attributes.h>
+#  include <cuda/std/__cccl/compiler.h>
+#  include <cuda/std/__cccl/dialect.h>
+#  include <cuda/std/__cccl/execution_space.h>
+#  include <cuda/std/type_traits>
+
+_LIBCUDACXX_BEGIN_NAMESPACE_CUDA
+
+enum class EvictionPolicy
+{
+  Default,
+  First,
+  Normal,
+  Last,
+  LastUse,
+  NoAllocation
+};
+
+enum class PrefetchSize
+{
+  NoPrefetch,
+  Bytes64,
+  Bytes128,
+  Bytes256
+};
+
+template <class _ElementType,
+          EvictionPolicy _Eviction = EvictionPolicy::Default,
+          PrefetchSize _Prefetch   = PrefetchSize::NoPrefetch,
+          typename _Enable         = void>
+struct cache_policy_accessor;
+
+/***********************************************************************************************************************
+ * accessor_reference
+ **********************************************************************************************************************/
+
+template <class _ElementType, EvictionPolicy _Eviction, PrefetchSize _Prefetch>
+class accessor_reference
+{
+  using pointer_type = _ElementType*;
+
+  pointer_type __p;
+
+  friend class cache_policy_accessor<_ElementType, _Eviction, _Prefetch>;
+
+public:
+  accessor_reference() = delete;
+
+  accessor_reference(accessor_reference&&) = delete;
+
+  _CCCL_HIDE_FROM_ABI _CCCL_DEVICE _CCCL_FORCEINLINE accessor_reference(const accessor_reference&) = default;
```

Comment:
I am not a big fan of putting `_CCCL_FORCEINLINE` everywhere

I was investigating making it part of `_CCCL_HIDE_FROM_ABI` but that lead to a ton of compiler issues

Reply:
I think it makes sense for small functions, especially in CUDA